trade-finance-manager-ui/server/controllers/case/cancellation/bank-request-date.controller.ts

```diff
@@ -37,6 +38,10 @@ export const getBankRequestDate = async (req: GetBankRequestDateRequest, res: Re
 
     const cancellation = await api.getDealCancellation(_id, userToken);
 
+    if (isEmpty(cancellation)) {
```

Comment:
What do you think about creating some middleware instead of having to do this in individual controllers?

The middleware could be for the `/cancellation/*` route with a condition so that if it's the "first" route we just call `next()`? I think it'd be a lot simpler

Reply:
There's currently no existing middleware that calls apis but I agree this could be really nice. 

I think this would need some thought about what pattern we'd want to follow & what benefits we'd gain - we could have a call sometime to chat about this but I think its outside the scope of this PR

Reply:
Sorry for the long comment - I think there'd be a few benefits:

- None of the controllers need to make the same API call
- One middleware handles it all, DRY
- If we need to update the redirect or add/change some logic, can just update in one place

I think there's only 1/2 ways to create the middleware, e.g

```ts
const canCancelDealCancellation = async (req, res, next) => {
  if (isRelevantRoute()) {
    try {
      const { _id } = req.params;

      const { user, userToken } = asUserSession(req.session);

      const cancellation = await api.getDealCancellation(_id, userToken);

      if (isEmpty(cancellation)) {
        return res.redirect(`/case/${_id}/deal`);
      }

      return next();
    } catch () {
      // ... 
    }
  }

  return next();
}
```

But this made me think - the controllers also call `api.getDealCancellation` to get a cancellation object. This would be required in the controllers and the middleware. So we could also create a second middleware that gets called first and stores the cancellation object in `res.locals` e.g

```ts
const getDealCancellation = async () => {
  if (isRelevantRoute()) {
    const cancellation = await api.getDealCancellation(_id, userToken);

    res.locals.dealCancellationData = cancellation;
  }

  return next();
};
```


This way we're doing the "get cancellation data" and "redirect if need to" type logic in one single place.

What do you think - shall we create a ticket for this? Happy to get this PR without such middleware.

Reply:
Strange one - feels like we're kind of introducing some tech debt rather than doing it straight away, but sometimes the implementation can highlight other ways to do it.

Happy to do this in another ticket. Can we create a JIRA ticket please?

Reply:
I think this would be a pretty large change of pattern - no other middleware makes api-requests. I've tried to find any documentation talking about this &[ Next.js explicity advise against this](https://nextjs.org/docs/app/building-your-application/routing/middleware)
> Middleware is not designed for direct data fetching or manipulation, this should be done within Route Handlers or server-side utilities instead.

Moreover, this would require quite a few pieces of new middleware
- Get the deal, return 404 if it doesnt exist, add it to the `res`
- `validateDealCanBeCancelled` Validate the deal is MIN/AIN
- get the cancellation & add it to the `res`
- `checkCancellationNotEmpty` redirects if the deal is empty

If we do want to add a new pattern, I'd want to arrange a meeting with the other devs to decide what the scope of making this change is & how much dev time we'd want to commit to it.

Reply:
Agree with everything but the first point @MarRobSoftwire that's for NextJS, we're not using NextJS.

Fetching data in NodeJS middleware is very common - for example for user authentication.

Having the same kind of functions in routes - being called before controllers, is pretty much doing the same thing.